src/ol_infrastructure/applications/jupyterhub/deployment.py

```diff
@@ -0,0 +1,436 @@
+"""Function for provisioning a JupyterHub deployment
+with values consistently derived from its name.
+"""
+
+from pathlib import Path
+
+import pulumi_kubernetes as kubernetes
+import pulumi_vault as vault
+from pulumi import Config, ResourceOptions, StackReference
+
+from bridge.lib.magic_numbers import DEFAULT_POSTGRES_PORT
+from bridge.lib.versions import JUPYTERHUB_CHART_VERSION
+from ol_infrastructure.components.aws.database import OLAmazonDB, OLPostgresDBConfig
+from ol_infrastructure.components.services.cert_manager import (
+    OLCertManagerCert,
+    OLCertManagerCertConfig,
+)
+from ol_infrastructure.components.services.k8s import (
+    OLApisixOIDCConfig,
+    OLApisixOIDCResources,
+    OLApisixRoute,
+    OLApisixRouteConfig,
+    OLApisixSharedPlugins,
+    OLApisixSharedPluginsConfig,
+)
+from ol_infrastructure.components.services.vault import (
+    OLVaultDatabaseBackend,
+    OLVaultK8SDynamicSecretConfig,
+    OLVaultK8SResources,
+    OLVaultK8SResourcesConfig,
+    OLVaultK8SSecret,
+    OLVaultPostgresDatabaseConfig,
+)
+from ol_infrastructure.lib.ol_types import StackInfo
+from ol_infrastructure.lib.stack_defaults import defaults
+from ol_infrastructure.lib.vault import postgres_role_statements
+
+
+def provision_jupyterhub_deployment(  # noqa: PLR0913
+    stack_info: StackInfo,
+    jupyterhub_deployment_config: Config,
+    vault_config: Config,
+    db_config: OLPostgresDBConfig,
+    cluster_stack: StackReference,
+    app_db: OLAmazonDB,
+    application_labels: dict[str, str],
+    k8s_global_labels: dict[str, str],
+    extra_images: dict[str, dict[str, str]] | None = None,
+    menu_override_json: str | None = None,
+    disabled_extensions_json: str | None = None,
+    extra_config: str | None = None,
+) -> kubernetes.helm.v3.Release:
+    """Provision a JupyterHub deployment with all required AWS and Kubernetes resources.
+
+    Args:
+        base_name: Base name for the deployment (e.g., "jupyterhub" or
+            "jupyterhub-authoring"). This name is used to derive all resource
+            names throughout the deployment.
+        domain_name: Domain name for the deployment (e.g., "jupyter.mitlearn.mit.edu")
+        namespace: Kubernetes namespace for the deployment
+        stack_info: Stack information from parse_stack()
+        jupyterhub_deployment_config: Pulumi config for jupyterhub deployment
+        vault_config: Pulumi config for vault
+        cluster_stack: EKS cluster stack reference
+        application_labels: Labels to apply to Kubernetes resources
+        k8s_global_labels: Global Kubernetes labels
+        extra_images: Dictionary mapping image names to image configuration dictionaries (optional)
+        menu_override_json: JSON contents for menu override Jupyter config
+        disabled_extensions_json: JSON contents for disabled extensions Jupyter config
+        extra_config: Extra configuration values to merge into the Helm chart values
```

Comment:
This docstring isn't grounded in reality. I'd just delete it or have an AI fix it. IMO the args are descriptive enough on their own.